toolchain/parse/state.def

```diff
@@ -1052,6 +1057,44 @@ CARBON_PARSE_STATE_VARIANTS3(TypeAfterParams, Class, Interface, NamedConstraint)
 //   (state done)
 CARBON_PARSE_STATE(BaseDecl)
 
+// Handles processing of an `impl...as` declaration after the introducer.
+//
+// impl forall [ ...
```

Comment:
I think more of these need to be tested. I'm seeing `impl X as Y` but no other syntax variations.